File: stringCenter/middleware/tab.js
function Tab(){
  this.info = "";
  this.measureCount = 1;
  this.measures = [];
}

function Measure(info, stringCount, tuning){
  this.info = info;
  this.stringCount = stringCount;
  this.strings = [];
  this.tuning = tuning;
}

function InstString(tuning, noteCount){
  this.tuning = tuning;
  this.noteCount = noteCount;
  this.notes = [];
}


Tab.prototype.initTab = function(info, stringCount, tuning, noteCount){
  this.addMeasure(info, stringCount, tuning, noteCount);
}

Tab.prototype.addMeasure = function(info, stringCount, tuning, noteCount){
  this.measures.push(new Measure(info, stringCount));
  this.measures[this.measures.length - 1].initMeasure(info, stringCount, tuning, noteCount);
  this.measureCount = this.measures.length;
}

Tab.prototype.addNote = function(measureNum, note, noteNum, stringNum){
  this.measures[measureNum].strings[stringNum].notes[noteNum] = note;
}

Tab.prototype.displayTab = function(){
  for(var i = 0; i < this.getMeasureCount(); i++){
    console.log("Measure " + i + ": " + this.measures[i].info);
    for(var j = 0; j < this.getStringCount(); j++){
<<<<<<< HEAD
      var instStringToBePrinted = this.measures[i].strings[j].tuning;
=======
      var instStringToBePrinted = this.measures[i].strings[j].tuning + " ";
>>>>>>> e7ff19ee845a186546317f14f1e627b390968570
      for(var k = 0; k < this.getNoteCountPerString(); k++){
        instStringToBePrinted += this.measures[i].strings[j].notes[k];
      }
      console.log(instStringToBePrinted);
    }
    console.log("");
  }
}

Tab.prototype.getStringCount = function(){
  return this.measures[0].stringCount;
}

Tab.prototype.getMeasureCount = function(){
  return this.measureCount;
}

Tab.prototype.getNoteCountPerString = function(){
  return this.measures[0].strings[0].noteCount;
}

Tab.prototype.getTuning = function(){
  return this.measures[0].tuning;
}

Measure.prototype.initMeasure = function(info, stringCount, tuning, noteCount){
  this.info = info;
  this.tuning = tuning;
  for(var i = 0; i < stringCount; i++){
    this.strings.push(new InstString(tuning[(tuning.length - 1) - i], noteCount));
    this.strings[i].initString(noteCount);
  }
}

InstString.prototype.initString = function(noteCount){
  for(var i = 0; i < noteCount; i++){
    this.notes.push("-");
  }
}

var tab = new Tab();

module.exports = tab;

File: tab.dart
import 'dart:convert';
import 'dart:io';


class Tabb{
  String _info;
  String _tuning;
  num _stringCount;
  List<Measure> _measures;

  Tabb(String info, String tuning){
    _info = info;
    _measures = [new Measure("first measure", tuning.length, 4, tuning)];
    _tuning = tuning;
    _stringCount = tuning.length;
  }

  Tabb.fromJson(encodedJson){
    Map json = JSON.decode(encodedJson);
    _info = json['info'];
    _tuning = json['tuning'];
    num measureCount = json['measureCount'];
    _measures = [];
    for(int i = 0; i < measureCount; i++) {
      _measures.add(new Measure.fromJson(json['measures'][i]));
    }
  }

  Map toJson(){
    Map map = new Map();
    map['info'] = _info;
    map['tuning'];
    map['measureCount'] = _measures.length;
    map['measures'] = _measures;
    return map;
  }

  void addMeasure(String info, int noteCount){
    _measures.add(new Measure(info, _stringCount, noteCount, _tuning));
  }

  void printTabb(){
    String temp = "";
    for(Measure measure in _measures){
      print(measure.info);
      for(InstString string in measure.strings){
        temp += string.tuning + ' ';
        for(String note in string.notes){
          temp += note;
        }
        print(temp);
        temp = "";
      }
    }
  }

  get info => _info;
  get tuning => _tuning;
  get measureCount => _measures.length;
}


class Measure{
  String _info;
  String _tuning;
  List<InstString> _strings;

  Measure(String info, int stringCount, int noteCount, String tuning){
    _info = info;
    _tuning = tuning;
    _strings = [];
    for(int i = 0; i < stringCount; i++){
      _strings.add(new InstString(noteCount, tuning[i]));
    }
  }

  Measure.fromJson(json){
    _info = json['info'];
    _tuning = json['tuning'];
    num stringCount = json['stringCount'];
    _strings = [];
    for(int i = 0; i < stringCount; i++){
      _strings.add(new InstString.fromJson(json['strings'][i]));
    }
  }

  Map toJson(){
    Map map = new Map();
    map['info'] = _info;
    map['tuning'] = _tuning;
    map['stringCount'] = _strings.length;
    map['strings'] = _strings;
    return map;
  }

  void addString(int noteCount, String tuning){
    _strings.add(new InstString(noteCount, tuning));
  }

  get info => _info;
  get tuning => _tuning;
  get stringCount => _strings.length;
  get strings => _strings;
}


class InstString  {
  List<String> _notes;
  String _tuning;

  InstString(int noteCount, String tuning){
    _notes = [];
    _tuning = tuning;
    for(int i = 0; i < noteCount; i++){
      _notes.add('-');
    }
  }

  InstString.fromJson(json){
    _notes = json['notes'];
    _tuning = json['tuning'];
  }

  void addNote(){
    _notes.add('-');
  }

  Map toJson(){
    Map map = new Map();
    map['notes'] = _notes;
    map['tuning'] = _tuning;
    return map;
  }

  get tuning => _tuning;
  get noteCount => _notes.length;
  get notes => _notes;
}